src/aiogram_dialog/context/intent_middleware.py

```diff
@@ -74,16 +74,18 @@ async def _load_stack(
     ) -> Optional[Stack]:
         if stack_id is None:
             raise InvalidStackIdError("Both stack id and intent id are None")
-        stack = await proxy.load_stack(stack_id)
-        if not await self.access_validator.is_allowed(stack, event, data):
-            user = data["event_from_user"]
-            logger.debug(
-                "Stack %s is not allowed for user %s",
-                stack.id, user.id,
-            )
+        try:
+            stack = await proxy.load_stack(stack_id)
+            if not await self.access_validator.is_allowed(stack, event, data):
+                user = data["event_from_user"]
+                logger.debug(
+                    "Stack %s is not allowed for user %s",
+                    stack.id, user.id,
+                )
+                return
+            return stack
+        finally:
```

Comment:
the idea was no unlock only if stack is not allowed. 
otherwise it is unlocked in the end of event handling